granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -78,7 +101,94 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
     return path
 
 
-def get_cgroup_mount(controller: str, resolve_host_root_links: bool = True) -> Optional[str]:
+CGROUP_PROCS_FILE = "cgroup.procs"
+
+
+class CgroupCore:
+    path: Path
+
+    def __init__(self, path: Path):
+        self.path = path
+
+    def _get_parent_cgroup(self, cgroup_name: str) -> Optional[Path]:
+        if cgroup_name in self.path.parts:
+            return Path(self.path.as_posix().split(cgroup_name)[0]) / cgroup_name
+        return None
+
+    def _get_cgroup_in_hierarchy(self, cgroup_name: str) -> Optional[Path]:
+        if self.has_parent_cgroup(cgroup_name):
+            new_cgroup_path = self._get_parent_cgroup(cgroup_name)
+        else:
+            new_cgroup_path = self.path / cgroup_name
+            new_cgroup_path.mkdir(exist_ok=True)
+
+        return new_cgroup_path
+
+    def has_parent_cgroup(self, cgroup_name: str) -> bool:
+        """
+        Check if cgroup has the given cgroup_name in its hierarchy
+        :param cgroup_name: the cgroup name
+        :return: bool indicating if cgroup is in hierarchy
+        """
+        return self._get_parent_cgroup(cgroup_name) is not None
+
+    def get_pids_in_cgroup(self) -> set[int]:
+        return {int(proc) for proc in self.read_from_interface_file(CGROUP_PROCS_FILE).split()}
+
+    def assign_process_to_cgroup(self, pid: int = 0) -> None:
+        self.write_to_interface_file(CGROUP_PROCS_FILE, str(pid))
+
+    def read_from_interface_file(self, interface_name: str) -> str:
+        interface_path = self.path / interface_name
+        return interface_path.read_text()
+
+    def write_to_interface_file(self, interface_name: str, data: str) -> None:
+        interface_path = self.path / interface_name
+        interface_path.write_text(data)
+
+    @abstractmethod
+    def get_cgroup_in_hierarchy(self, controller: str, cgroup_name: str) -> CgroupCore:
+        """
+        Return a CGroup in the hierarchy which has the given name.
+        If one doesn't exist higher up in the hierarchy, create a new one lower in the hierarchy.
+        """
+        pass
+
+
+class CgroupCoreV1(CgroupCore):
+    def get_cgroup_in_hierarchy(self, controller: str, cgroup_name: str) -> CgroupCore:
+        subcgroup_path = self._get_cgroup_in_hierarchy(cgroup_name)
+        assert subcgroup_path is not None
+        return CgroupCoreV1(subcgroup_path)
+
+
+CGROUP_V2_SUPPORTED_CONTROLLERS = "cgroup.controllers"
+CGROUP_V2_ENABLED_CONTROLLERS = "cgroup.subtree_control"
+
+
+class CgroupCoreV2(CgroupCore):
+    def is_controller_supported(self, controller: str):
+        return controller in self.read_from_interface_file(CGROUP_V2_SUPPORTED_CONTROLLERS).split()
+
+    def enable_controller(self, controller: str):
+        assert self.is_controller_supported(controller), f"Controller not supported {controller!r}"
```

Comment:
To my understand, here you're protecting from this case:
![image](https://user-images.githubusercontent.com/8831572/215291222-83cbfeac-d521-4fa6-a0e6-5023a71e2cc0.png)

Please, instead of trying to mimic the protection logic that cgroup employs in the kernel, you just try to do things & catch the related errors and handle them per their meaning. In this case, you're `assert`ing which means you don't expect this to happen - so instead you can `try` and if you get `ENOENT` you narrow it down to `ENOENT` from the `write` call itself (perhaps separate to `open` and `write`) and THEN you raise an exception which tells the controller is disabled.

Reply:
Same as I said here https://github.com/Granulate/granulate-utils/pull/117#discussion_r1066443110

Reply:
No, this is a bug. I should've tested that `cgroup.subtree_control` of the current cgroup has the controller we want. I'm changing it to raise an exception if the controller is not delegated as we can't create a subcgroup in this case.

Reply:
https://github.com/Granulate/granulate-utils/pull/117#discussion_r1136654675 This comment is relevant here as well

Reply:
Resolving as this is no longer relevant